### PR TITLE
40 create client class

### LINKED ISSUE #40: Client handling
Introduce a client handler to store additional information about clients alongside the socket reference such as
ClientId - each client likely has a unique identifier
Subscribed topics - clients can subscribe to different channels, keeping track of these subscriptions
Status information - info about clients, connection, status, activity, and permissions etc

### PR DESCRIPTION
**Pull request by Emmelie83**
### Related Issue(s)
Closes #40

### Proposed Changes
Add a client class to the project and update the client handler class accordingly

- **Change 1**: Adding a client class 
- **Change 2**: Adding a topic handler class needed for the client class
- **Change 3**: Uppdating the client handler class accordingly
- **Change 3**: Changing the name of the server class to "Broker"

### Description
Added a client class, a topic handler class and updating the client handler accordingly. Added test for client class and topic handler class. Changed the name of the server class to "Broker".

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit 6f6387a050af28928459a63c9cd51b561121937d.  | 
|--------|--------|

### Summary:
This PR introduces new `Client` and `TopicHandler` classes, updates the `ClientHandler` class, renames the `Server` class to `Broker`, and includes relevant tests.

**Key points**:
- Introduced a `Client` class with methods for connecting, disconnecting, subscribing, and unsubscribing.
- Added a `TopicHandler` class to manage subscriptions.
- Updated the `ClientHandler` class to handle `Client` and `TopicHandler` classes.
- Renamed the `Server` class to `Broker`.
- Included tests for the new `Client`, `TopicHandler`, and `Broker` classes.
- Updated existing tests.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: src/test/java/org/fungover/thunder/ClientHandlerTest.java
Comment (cybertobbe) [inline src/test/java/org/fungover/thunder/ClientHandlerTest.java:1]: Missing test in the ClientHandlerTest?

File: src/main/java/org/fungover/thunder/Client.java
Comment (IngridHammargren) [inline src/main/java/org/fungover/thunder/Client.java:1]: Well done Emelie! I think your code looks good. Your class depend on our "Topic" class, so we need to work on making your classes applicable. It seems like your branch has a conflict, i don´t think I saw it before your latest commit.

Comment (Emmelie83) [reply src/main/java/org/fungover/thunder/Client.java:1]: Thanks! Yes, there was a merge to main that made my branch full of conflicts, but I have resolved them now :-)

Comment (Emmelie83) [reply src/main/java/org/fungover/thunder/Client.java:1]: I have a string as topic, but I see that you have created a class called Topic. The client class then probably should have a set of topics instead of strings then. So private Set<String> subscribedTopics will probably have to be changed to private Set<Topic> subscribedTopics.

**Comment by Emmelie83** [comment]
I am unsure if the client class should also have a publish method.

**Comment by vonfredd** [comment]
> I am unsure if the client class should also have a publish method.

> I am unsure if the client class should also have a publish method.

#25 #26 #27 Are issued features to handle Subscribe, Unsubscribe and Publish. I do not think you need to implement as much as possible now since there is a risk that everything will be very tightly coupled. I believe the features will be implemented when the issues are worked on 😄

File: src/main/java/org/fungover/thunder/Client.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    public void connect() {
+        if (!connected) {
+            // Establish a connection to the MQTT broker
+            // Perform the MQTT connection handshake
+            System.out.println("Client " + clientId + " connected to broker.");
+            connected = true;
+        } else {
+            System.out.println("Client " + clientId + " is already connected.");
+        }
+    }
```
Comment (vonfredd) [inline src/main/java/org/fungover/thunder/Client.java:36]: I do not really understand this. If a client is not connected, doesn't removeDisconnectedClients() in ClientHandler remove them from the list? Which would make this unnecessary?

File: src/main/java/org/fungover/thunder/Main.java [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
-        }
+    public static void main(String[] args) throws IOException {
+        ServerSocket serverSocket = new ServerSocket(1883);
+        ClientHandler clientHandler = new ClientHandler();
+        Broker broker = new Broker(clientHandler, serverSocket);
+        createAndStartBroker(broker);
     }
+
+    public static void createAndStartBroker(Broker broker) throws IOException {
+        broker.start();
```
Comment (vonfredd) [inline src/main/java/org/fungover/thunder/Main.java:16]: I do not understand this refactor. I feel it became more complex than before but seem to use same logic?

**Review by vonfredd — COMMENTED** [review]
There are some things I wonder about 😄

File: src/main/java/org/fungover/thunder/Client.java (hunk 1 above)
Comment (Emmelie83) [reply src/main/java/org/fungover/thunder/Client.java:36]: This is for checking whethere the client is connected or not. Just because it isn't connected, it doesn't disappear -- but it does get removed from the list of connected clients in the ClientHandler class. At least I thougt it works like this, but my networking skills aren't exactly great, I am kind of struggeling :-)

File: src/main/java/org/fungover/thunder/Main.java (hunk 2 above)
Comment (Emmelie83) [reply src/main/java/org/fungover/thunder/Main.java:16]: I truly understand that. The only reason I refactored was to be able to test the method and get the test coverage up to 80 % 😄😄😄😄

**Review by Danback — DISMISSED** [review]
Looks good to me, but the minor conflict needs to be solved before merging.

File: src/main/java/org/fungover/thunder/Client.java (hunk 1 above)
Comment (altmannk) [reply src/main/java/org/fungover/thunder/Client.java:36]: Makes sense, this one helps to know when the client is ready to receive data(that is connected or not) and the other one is to keep an accurate list of active clients, right? Do we have any code yet that monitor the clients status? Thinking of reconnection, I have a vague memory of Martin mentioning something about if a user is out in the woods and looses connection to the broker, that the system can try to reconnect automatically. 🤔

Comment (vonfredd) [reply src/main/java/org/fungover/thunder/Client.java:36]: > This is for checking whethere the client is connected or not. Just because it isn't connected, it doesn't disappear -- but it does get removed from the list of connected clients in the ClientHandler class. At least I thougt it works like this, but my networking skills aren't exactly great, I am kind of struggeling :-)

I wasnt thinking that we could save disconnected clients so I misunderstood the usage. Well done emmelie.

**Review by Danback — APPROVED** [review]

File: src/main/java/org/fungover/thunder/Main.java [hunk 3]
```diff
@@ -1,16 +1,15 @@
...
-            server = new Server();
-            server.start();
-        } catch (IOException | NullPointerException e) {
-            e.printStackTrace();
-        }
+    public static void main(String[] args) throws IOException {
+        ServerSocket serverSocket = new ServerSocket(1883);
+        ClientHandler clientHandler = new ClientHandler();
+        Broker broker = new Broker(clientHandler, serverSocket);
+        broker.start()
```
Comment (altmannk) [inline src/main/java/org/fungover/thunder/Main.java]: Missing a semicolon ';' at the end of broker.start()

**Review by altmannk — COMMENTED** [review]
Insert ';'

File: src/main/java/org/fungover/thunder/Main.java (hunk 3 above)
Comment (altmannk) [reply src/main/java/org/fungover/thunder/Main.java]: Never mind 😂

Comment (Emmelie83) [reply src/main/java/org/fungover/thunder/Main.java]: Thanks, I have now :-)

File: src/main/java/org/fungover/thunder/TopicHandler.java
Comment (marcusju23) [inline src/main/java/org/fungover/thunder/TopicHandler.java:1]: New PR appeared with Subscription class #58 . Maybe it's safe to remove this file?

**Review by marcusju23 — CHANGES_REQUESTED** [review]
Overall it looks great. However, it may be necessary to remove TopicHandler.java and TopicHandlerTest.java considering PR #58 .

**Comment by Emmelie83** [comment]
> Overall it looks great. However, it may be necessary to remove TopicHandler.java and TopicHandlerTest.java considering PR #58 .

Since that is a much later issue than this one, I think that should be dealt with in that issue. This pull request has been waiting for very long now.

**Comment by kappsegla** [comment]
> > Overall it looks great. However, it may be necessary to remove TopicHandler.java and TopicHandlerTest.java considering PR #58 .
> 
> Since that is a much later issue than this one, I think that should be dealt with in that issue. This pull request has been waiting for very long now.

Agree. So the best way forward now is to make sure this PR gets any outstanding review code changes handled and merged so other programmers don't start to implement same code. If a PR hangs around too long it might get outdated and not useful anymore and it might be better to close it. :)

File: src/main/java/org/fungover/thunder/Broker.java [hunk 4]
```diff
@@ -4,13 +4,13 @@
 import java.net.ServerSocket;
 import java.net.Socket;
 
-public class Server {
+public class Broker {
```
Comment (kappsegla) [inline src/main/java/org/fungover/thunder/Broker.java:7]: We could make this its own PR or are we sure other coders can handle this seamless so we don't end up with two classes, both a Server and a Broker class? Is this really a Broker and is it inside the scope of the PR?

File: src/main/java/org/fungover/thunder/ClientHandler.java [hunk 5]
```diff
@@ -31,8 +32,8 @@ public void handleConnection(Socket clientSocket) {
                 }
             }
             removeDisconnectedClients();
-        } catch (Exception e) {
-            e.printStackTrace();
+        } catch (IOException e) {
```
Comment (kappsegla) [inline src/main/java/org/fungover/thunder/ClientHandler.java:35]: Is this needed or just for refactoring?

File: src/main/java/org/fungover/thunder/Main.java
Comment (kappsegla) [inline src/main/java/org/fungover/thunder/Main.java:1]: This might be a good refactor but out of scope for this PR? So lets do that in it's own PR

File: src/main/java/org/fungover/thunder/TopicHandler.java
Comment (kappsegla) [inline src/main/java/org/fungover/thunder/TopicHandler.java:1]: Is this part of the PR? Wasn't it to create a Client class that can store information about the connected client? :)

File: src/main/java/org/fungover/thunder/Broker.java (hunk 4 above)
Comment (Emmelie83) [reply src/main/java/org/fungover/thunder/Broker.java:7]: There was not much to change as I did the PR, but now it could be more problematic to make this change. So probably I should dismiss my changes. I have lately also thougt that is could as well be called Server :-)

File: src/main/java/org/fungover/thunder/ClientHandler.java (hunk 5 above)
Comment (kappsegla) [reply src/main/java/org/fungover/thunder/ClientHandler.java:35]: Yes, Sonar doesn't like printing the stacktrace for security reasons since we then might expose what methods and classes are called. 
The right way of solving it is to replace with a log message using java.util.logger.
Add this as a field:
    private static final Logger logger = Logger.getLogger(Broker.class);
    
   And then you can log an error using logger object.

File: src/main/java/org/fungover/thunder/Client.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package org.fungover.thunder;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class Client {
+    private final String clientId;
+    private boolean connected;
+    private Set<String> subscribedTopics;
```
Comment (kappsegla) [inline src/main/java/org/fungover/thunder/Client.java:9]: This looks good. 👍

File: src/main/java/org/fungover/thunder/Client.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    public void connect() {
+        if (!connected) {
+            // Establish a connection to the MQTT server
+            // Perform the MQTT connection handshake
+            System.out.println("Client " + clientId + " connected to server.");
+            connected = true;
+        } else {
+            System.out.println("Client " + clientId + " is already connected.");
+        }
+    }
```
Comment (kappsegla) [inline src/main/java/org/fungover/thunder/Client.java:34]: Client in our application means a class that we can create objects of that represents a connected client. So it isn't a client that connects to a mqtt broker/server. So when calling connect just mark this client as connected.
Logging of what happens might be better placed in the code that accepts new connections :)

**Comment by Emmelie83** [comment]
Not enough Test coverage. But I shouldn't test the Main method, should I? The Client Class has 100 %, do I don't really know what to do about it.

**Review by kappsegla — APPROVED** [review]

**Comment by kappsegla** [comment]
> Not enough Test coverage. But I shouldn't test the Main method, should I? The Client Class has 100 %, do I don't really know what to do about it.

Lets just ignore this since it's hard to test that part of the code with unit tests. Better to do integration testing and start a real server for that. Good work!

File: src/test/java/org/fungover/thunder/ClientTest.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    @Test
+    public void givenClientHasNotSubscribedToAnyTopicThenGetSubscribedTopicsShouldReturnAnEmptyHashSet() {
+        String clientId = "testClient";
+        Client client = new Client(clientId);
+
+        assertTrue(client.getSubscribedTopics().isEmpty());
+    }
+
+    @Test
+    public void givenClientHasSubscribedToATopicTwiceThenGetSubscribedTopicsShouldReturnASetWithOnlyOneOccurrenceOfTheTopic() {
```
Comment (altmannk) [inline src/test/java/org/fungover/thunder/ClientTest.java:50]: Less code to read if we skip public modifier before test methods. :)

**Review by altmannk — CHANGES_REQUESTED** [review]
Went through your code, it's looking good. Just a minor touchup in the ClientTest, then it's good to go! 🚀

File: src/main/java/org/fungover/thunder/Client.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+public class Client {
+    private final String clientId;
+    private boolean connected;
+    private Set<String> subscribedTopics;
+
+    public Client(String clientId) {
+        this.clientId = clientId;
+        this.connected = false;
+        this.subscribedTopics = new HashSet<>();
+    }
```
Comment (vonfredd) [inline src/main/java/org/fungover/thunder/Client.java:15]:
```suggestion
public class Client {
    private final String clientId;
    private boolean connected;
    private Set<String> subscribedTopics;
    private Socket socket;

    public Client(String clientId) {
        this.clientId = clientId;
        this.connected = false;
        this.subscribedTopics = new HashSet<>();
        this.socket = new Socket();
    }
```
One thing I am missing is the reference to a socket so we can read from and send to the client. Other than that I think it looks good.

File: src/test/java/org/fungover/thunder/ClientTest.java (hunk 8 above)
Comment (vonfredd) [reply src/test/java/org/fungover/thunder/ClientTest.java:50]: I also think the public modifiers should be removed.

**Review by vonfredd — CHANGES_REQUESTED** [review]
I think it is done after removing public modifier for the test methods and give the client a Socket field.

**Review by altmannk — APPROVED** [review]

**Review by kappsegla — APPROVED** [review]

**Review by marcusju23 — APPROVED** [review]

**Review by vonfredd — APPROVED** [review]
